### PR TITLE
support categorical values.

### PR DESCRIPTION
**Pull request by zheng-da**
*Description of changes:*
We can convert some values into categorical values.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/transform.py [hunk 1]
```diff
@@ -97,6 +101,74 @@ def collect_info(self, info):
...
+        Parameters
+        ----------
+        feats : np array
+            Data with categorical values.
+
+        Returns
+        -------
+        np.array
+        """
+        return {self.feat_name: np.array([self._val_dict[val] for val in feats])}
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:152]: This code only handle the case when each node belongs to only one category.

File: python/graphstorm/gconstruct/transform.py [hunk 2]
```diff
@@ -97,6 +101,74 @@ def collect_info(self, info):
...
+
+    def pre_process(self, feats):
+        """ Pre-process data
+
+        Parameters
+        ----------
+        feats: np.array
+            Data to be processed
+        """
+        return {self.feat_name: np.unique(feats)}
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:128]: numpy support strings. my unit test uses strings

File: python/graphstorm/gconstruct/transform.py (hunk 1 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:152]: this is to handle categorical values. to handle multiple categorical values, we need to define how multiple categorical values are stored.

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -97,6 +101,91 @@ def collect_info(self, info):
...
+        if self._separator is None:
+            for i, feat in enumerate(feats):
+                encoding[i, self._val_dict[feat]] = 1
+        else:
+            for i, feat in enumerate(feats):
+                idx = [self._val_dict[val] for val in feat.split(self._separator)]
+                encoding[i, idx] = 1
+        return {self.feat_name: encoding}
+
+    def save(self, file_path):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:171]: Where does save() be called?

File: python/graphstorm/gconstruct/transform.py [hunk 4]
```diff
@@ -97,6 +101,91 @@ def collect_info(self, info):
...
+        ----------
+        feats: np.array
+            Data to be processed
+        """
+        if self._separator is None:
+            return {self.feat_name: np.unique(feats)}
+        else:
+            vals = []
+            for feat in feats:
+                vals.extend(feat.split(self._separator))
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:136]: why set is efficient than list?

File: python/graphstorm/gconstruct/transform.py (hunk 3 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:171]: it's not. we can remove it.

File: python/graphstorm/gconstruct/transform.py [hunk 5]
```diff
@@ -97,6 +101,91 @@ def collect_info(self, info):
...
+        Parameters
+        ----------
+        feats : np array
+            Data with categorical values.
+
+        Returns
+        -------
+        np.array
+        """
+        encoding = np.zeros((len(feats), len(self._val_dict)), dtype=np.int8)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/transform.py:161]: https://github.com/awslabs/graphstorm/blob/main/python/graphstorm/model/embed.py#L239
it converts data type automatically.

File: docs/source/configuration/configuration-gconstruction.rst [hunk 6]
```diff
@@ -83,6 +83,7 @@ Currently, the graph construction pipeline only supports three feature transform
...
 * **Numerical MAX_MIN transformation** normalizes numerical input features with $val = (val-min)/(max-min)$, where $val$ is the feature value, $max$ is the maximum number in the feature and $min$ is the minimum number in the feature. The ``name`` field in the feature transformation dictionary is ``max_min_norm``. The dict can contains two optional fields. ``max_bound`` specifies the maximum value allowed in the feature. Any number larger than ``max_bound`` will be set to ``max_bound``. ``min_bound`` specifies the minimum value allowed in the feature. Any number smaller than ``min_bound`` will be set to ``min_bound``.
+* **Convert to categorical values** converts text data to categorial values. The `name` field is `to_categorical`. `separator` specifies how to split the string into multiple categorical values (this is only used to define multiple categorical values). If `separator` is not specified, the entire string is a categorical value. `mapping` specifies how to map a string to an integer value that defines a categorical value.
```
Comment (classicsong) [inline docs/source/configuration/configuration-gconstruction.rst:86]: Do we need to provide the format of mapping here?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 7]
```diff
@@ -605,6 +612,8 @@ def process_graph(args):
     argparser = argparse.ArgumentParser("Preprocess graphs")
     argparser.add_argument("--conf-file", type=str, required=True,
                            help="The configuration file.")
+    argparser.add_argument("--output-conf-file", type=str,
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:615]: Default is None?

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 8]
```diff
@@ -327,6 +332,75 @@ def test_feat_ops():
...
+        0: preproc_res0,
+        1: preproc_res1
+    }
+    update_two_phase_feat_ops(return_dict, res7)
+    proc_res3 = process_features(data0, res7)
+    assert "test7" in proc_res3
+    for one_hot, str_i in zip(proc_res3["test7"], data0["test1"]):
+        assert one_hot[int(str_i)] == 1
+    assert 'mapping' in feat_op7[0]["transform"]
+
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:364]: Can we add a test as passing feat_op7[0]["transform"]['mapping'] to categorical transform and do the data transform again on data0 and data1 and campare the new `one_hot` with the one in L361?

**Review by classicsong — APPROVED** [review]
Overall LGTM.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 7 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:615]: by default, we don't output the new config file.

**Pull request merged**